site/drafts/withered.md

```diff
@@ -0,0 +1,431 @@
+---
+title: 'Composable filters using Witherable optics'
+author: "Chris Penner"
+date: "Oct 27, 2020"
+tags: [haskell, optics]
+description: ""
+---
+
+One of my favourite things about Haskell is that its structures and abstractions
+are very principled, and they have laws dictating correct behaviour.
+
+In my experience, this means that when you find a new way to piece together those
+abstractions it _almost always_ ends up doing something reasonable... or at the very least interesting!
+
+It comes as no surprise to some of you that I think about **optics** (lenses) an awful lot! 
+Optics are the most beautiful, composable, expressive, and adaptable abstractions I've
+ever had the pleasure of discovering. Optics provide an elegant abstraction
+for many data-manipulation strategies that have been around for decades.
+
+In this post we're going to explore a new type of optic which provides composable
+and expressive means for filtering complex data structures.
+To do so is going to require a base-level understanding of optics, and preferably readers would
+understand what a Traversal is. Bonus points if you've implemented one on your own or have some idea of how they work!
+
+## The Background
+
+Let's take a look at the Traversal typeclass, it's where we find Haskell's
+all-powerful secret weapon **traverse**! ([BTW The answer is always traverse.](https://impurepics.com/fp-bot/index.html))
+
+```haskell
+class (Functor t, Foldable t) => Traversable t where
+  traverse :: Applicative f => (a -> f b) -> t a -> f (t b)
+```
+
+Many beginners rightfully stumble on this function because it's just so dang abstract! It's tough to tell what it does for you, and how you could use it (the answer is almost anywhere).
+
+
+Eventually, folks realized that we could write more "specialized" versions of this type which worked on anything, not just Traversable datatypes, and the Traversal was born!
+
+A traversal is anything that fits this rough shape:
+
+```haskell
+forall f. Applicative f => (a -> f b) -> (s -> f t)
+```
+
+We can find different optics by specializing the constraints on `f` in interesting ways, for instance lenses require `f` to be a `Functor`, and folds require `f` to have both `Contravariant` and `Applicative` instances (exercise: go see if you can figure out what those two constraints imply!).
+
+Anyways, long story short, if you can make your function fit that shape, it's probably useful as some sort of optic!
+
+This leads us to Witherable!
+
+## Witherable
+
+Ever heard of [Witherable](https://hackage.haskell.org/package/witherable-0.3.5/docs/Data-Witherable.html)?
+It's a class which extends from Traversable, that is, Traversable is a superclass, all Witherables are Traversable, but not the other way around.
```

Comment:
You could mention here that `Traversable`s have a hidden invariant: they must preserve the size of a container in order to be lawful. `Witherable` is a subclass of `Traversable` precisely because you can optionally "shrink" (i.e. "wither") a container, but also reproduce the `Traversable` invariant via 

```haskell
traverse f ~ wither (fmap Just . f)
```